src/channel.c

```diff
@@ -1889,11 +1889,17 @@ ssize_t _libssh2_channel_read(LIBSSH2_CHANNEL *channel, int stream_id,
            return code. */
         if(channel->remote.eof || channel->remote.close)
             return 0;
-        else if(rc != LIBSSH2_ERROR_EAGAIN)
-            return 0;
 
-        /* if the transport layer said EAGAIN then we say so as well */
-        return _libssh2_error(session, rc, "would block");
+        /* If the transport layer said EAGAIN then we say so as well.
+           If there wasn't any error, it means the transport layer got
+           new data but it went to some other place and so we return
+           EAGAIN too. */
+        if((rc == LIBSSH2_ERROR_EAGAIN) || (rc == LIBSSH2_ERROR_NONE))
```

Comment:
After so long, I can not remember the details.

I think that I was getting spurious LIBSSH2_ERROR_NONE return codes from libssh2_channel_read, and I did supposed it was caused by data being eat at some upper layer.

But now, looking at the code (`libssh2_transport_read`), the only ways in which a LIBSSH2_ERROR_NONE can be returned are the two pointed by you and which should not happen under normal circumstances (or are servers sending packets of type 0?). So, maybe I was just trying to fix a bug that was not there at all.

Anyway, I still think that reserving 0 for signaling an EOF condition is a must. 

And the LIBSSH2_SOCKET_DISCONNECTED condition should also be handled properly, returning that error or 0 for EOF.

I would try to get back over this and recheck it in the next days.